doc/shared/sunadjoint/SUNAdjointStepper.rst

```diff
@@ -0,0 +1,327 @@
+   N_Vector sens, sunrealtype* tret)
+
+   Evolves the adjoint system backwards one step.
+
+   :param adj_stepper: The adjoint solver object.
+   :param tout: The time at which the adjoint solution is desired.
+   :param sens: The vector of sensitivity solutions :math:`\partial g/\partial y_0` and :math:`\partial g/\partial p`.
+   :param tret: On return, the time reached by the adjoint solver.
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+
+.. c:function:: SUNErrCode SUNAdjointStepper_RecomputeFwd(SUNAdjointStepper adj_stepper, suncountertype start_idx,\
+                                                          sunrealtype t0, sunrealtype tf, N_Vector y0)
+
+   Evolves the forward system in time from (``start_idx``, ``t0``) to (``stop_idx``, ``tf``) with dense checkpointing.
+
+   :param adj_stepper: The SUNAdjointStepper object.
+   :param start_idx: the index of the step, w.r.t. the original forward integration, to begin forward integration from.
+   :param t0: the initial time, w.r.t. the original forward integration, to start forward integration from.
+   :param tf: the final time, w.r.t. the original forward integration, to stop forward integration.
+   :param y0: the initial state, w.r.t. the original forward integration, to start forward integration.
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+
+.. c:function:: SUNErrCode SUNAdjointStepper_SetJacFn(SUNAdjointStepper adj_stepper, SUNRhsJacFn JacFn, \
+      SUNMatrix Jac, SUNRhsJacFn JacPFn, SUNMatrix JacP)
+
+   Sets the function pointers and matrices needed to evaluate and store :math:`\partial f / \partial y` and
+   :math:`\partial f / \partial p`. ``Jac`` should have dimensions ``neq x neq`` where ``neq`` is the number of states
+   in the forward problem. ``JacP`` should have dimensions ``nparams x neq`` where ``nparams`` is the
+   number of parameters in the model to get sensitivities for.
+
+   :param adj_stepper: The SUNAdjointStepper object.
+   :param JacFn: the function that evaluates :math:`\partial f / \partial y`.
+   :param Jac: a :c:type:`SUNMatrix` that will hold :math:`\partial f / \partial y`.
+   :param JacPFn: the function that evaluates :math:`\partial f / \partial p`.
+   :param JacP: a :c:type:`SUNMatrix` that will hold :math:`\partial f / \partial p`.
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
+
+
+.. c:function:: SUNErrCode SUNAdjointStepper_SetJacHermitianTransposeVecFn(SUNAdjointStepper adj_stepper, SUNRhsJacTimesFn Jvp, SUNRhsJacTimesFn JPvp)
+
+
+   Sets the function pointers to evaluate :math:`(\partial f/\partial y)^* v`  and :math:`(\partial f/\partial p)^* v`
+
+   :param adj_stepper: The SUNAdjointStepper object.
+   :param Jvp: function that evaluates :math:`(\partial f/\partial y)^* v`.
+   :param JPvp: function that evaluates :math:`(\partial f/\partial p)^* v`.
+
+   :return: A :c:type:`SUNErrCode` indicating failure or success.
```

Comment:
These `Set` functions are specific to a problem with one right-hand side function. To support future extensions, I think we should remove these and have the user pass right-hand side functions to `ERKStepCreateAdjointStepper` and `ARKStepCreateAdjointStepper` that compute `(\partial f/\partial y)^* v` and `(\partial f/\partial p)^* v`. This is probably clearer in the context of that PR so I'll comment there as well.

Reply:
This change has been made.